changes/2020-06-17_encrypt-max-plaintext-length-input/change.md

```diff
@@ -0,0 +1,205 @@
+
+This change renames this input to `plaintext length bound` in the specification
+to be a more accurate reflection of its behavior and intent,
+and specifies the correct behavior during the Encrypt operation.
+Specifically, in the scope of the Encrypt operation its value MUST be passed to the CMM.
+
+Additionally, it is unclear in the spec what the behavior should be
+if the input plaintext has a known-length
+and the customer specifies a bound on the length of that plaintext.
+
+This change proposes that in the scope of the Encrypt operation,
+an input plaintext is either of known length,
+of unknown length with customer supplied estimated size,
+or unknown length with no customer supplied estimated size.
+The Encrypt operation should never consider a `plaintext length bound` alongside
+a known-length plaintext.
+
+How a customer should express on the API level the intent to encrypt a known-length plaintext
+or unknown-length plaintext with or without a bound on the length
+depends on the specific implementation
+and the idioms of that implementation's language to describe APIs.
+
+## Out of Scope
+
+- Significantly changing the shape of any of our Encrypt API implementations is out of scope.
+
+## Motivation
+
+All implementations allow customers to specify a plaintext length when calling Encrypt with a
+unknown-length plaintext.
+However, its behavior in the Encrypt operation is not clearly specified.
+
+Additionally, some implementations also allow a plaintext length to be specified on Encrypt with a
+known-length plaintext.
+The behavior of Encrypt in this case is not consistent between implementations.
+
+The purpose of this change is to specify exactly how this input affects the Encrypt operation.
+
+This change renames this input to `plaintext length bound` within the spec because
+the intention behind the value of this input is that
+it is a max bound of the plaintext length,
+set by the customer when streaming encryption.
+If the actual plaintext length is greater than this value,
+something is wrong and the operation MUST fail.
+Otherwise, the Encrypt operation MUST pass this value to the CMM.
+
+The intention of this value is not to be a one to one passthrough value to the CMM.
+If that were the case,
+then we would be concerned with always letting customers specify this value for known-length plaintexts,
+or with possibly letting customers pass in a "unknown length" intent.
+We do not want to support these cases.
+Customers SHOULD NOT set a `max plaintext length` value for known-length plaintexts
+with the intention of changing CMM behavior.
+Instead, customers SHOULD specify this value for unknown-length plaintexts in order to use CMMs
+that depend on plaintext length for their internal logic.
+
+This change additionally renames the `plaintext length` field on GetEncryptionMaterialsRequest to
+`max plaintext length` because that better explains the intent behind this field.
```

Comment:
I've read this through several times now and I just get more confused each time. :/

What is the difference between `plaintext length bound` and `max plaintext length` other than one being part of the Encrypt API and one being part of the CMM GetEncryptionMaterials API?

Reply:
`max plaintext length` is an input on the CMM GetEncryptionMaterials API. It describes the maximum size of the plaintext that the CMM caller intends to encrypt using the result materials. I assert that in the context of Encrypt, if the plaintext has a known length this MUST be exactly the length of the plaintext.

`plaintext length bound` is an input on Encrypt. If provided, it describes the maximum size of a unknown-length plaintext. It specifically MUST NOT describe the intent of "I, as a caller, want the Encrypt operation to pass this value in to the CMM instead of the actual length."

They both describe "plaintext MUST NOT be larger than this size", but the intentions are subtly different. In order to distinguish them, I have given them slightly different names.